### ISSUE TITLE
Ignore punctuation when measuring 'words without misstrokes' 

### ISSUE THREAD

**Issue by jladdjr** [didoesdigital/typey-type#11]
When you click on the [progress page](https://didoesdigital.com/typey-type/progress), the top of the page reports `You’ve successfully typed <some number> words without misstrokes`. From what I can tell, the count here considers `foo`, `foo.`, `foo!`, etc. to be different "words". When I first started learning steno, learning those variations did seem significant, but having become more comfortable with punctuation, `foo` and `foo!` don't feel like they represent an increase in my vocabulary. I've noticed lately that I'm not paying much attention to this metric, and instead tracking the progress reported by the `Top 10000 Project Gutenberg words` lesson.

Recommend tweaking this metric to track words by themselves, instead of the current `word + <punctuation>` behavior. 

---

* In the description, I focused on the `successfully typed X words` metric, but I suspect this applies to the other metrics listed in that same area as well (so, `memorised words` and `revise` words, as well)
* I haven't done a ton of testing on this issue, this is mostly just based on what I've observed over time. I'm fairly sure that the metrics consider words with different punctuation to be distinct from each other. I also suspect that spacing may affect the counts as well.

**Comment by didoesdigital** [comment]
Thanks for adding this @jladdjr . I'll share some thoughts for discussion.

That's right, the following metrics all use the same definition of "words" for measuring progress:

1. X words without misstrokes
2. X memorised words
3. X seen words
4. X% of the way to 10,000 words
5. Today’s efforts: Old words
6. Today’s efforts: New words

"Words" here include all individual entries in the Progress JSON shown in "Words you’ve seen and times you’ve typed them well". That means that similar entries with different prefixes/suffixes, capitalisation, spacing, or punctuation are all considered different words. For example, these are all counted separately:

- `foo`
- ` foo`
- `foo `
- `foo,`
- ` foo,`
- `foo, `
- `Foo`
- `FOO`
- `foo bar`
- `foobar`

This can be useful in cases where you want to learn a word with and without punctuation separately. For example, `"PWA*": "bah",` and `"PWAERBGS": "bah{,}",`. A brief that includes punctuation might seem more meaningfully different as a "word" than comparing the word and standard punctuation strokes like `PWA*/KW-BG` for `bah,`.

This means the progress shown is an **accurate** measure but perhaps not as **obvious/learnable** or **satisfying**.

We could update the 6 Progress page word count metrics to add together the counts for entries containing the same words with and without punctuation. We would then also need to update the progress revision lessons setup to match.

We may or may not update the 6 Progress page word count metrics to add together the counts for entries with different leading/trailing spacing.

We would need to decide what punctuation counts as "the same". A naive approach might just trim trailing commas, full stops, exclamation marks, and question marks before adding its count to the word's count. 

A more sophisticated approach might need to consider:

- double quotes, which includes checking for escaping backslashes e.g. `" brains.\""`
- single quotes, where some words might still need to be counted separately e.g. `'em` vs `em`, `'round` vs `round`, `'cause` vs `cause`, and `s` vs `'s`.
- brackets of any kind e.g. `()[]{}<>⟨ ⟩< >⸤ ⸥｢ ｣⟦ ⟧〔 〕‹ ›« »`
- other common punctuation e.g. `[!"#$%&'()*,./:;<=>?@[\]^\`{|}~]-`
- leading punctuation e.g. `" --From"`
- leading/trailing pairs of punctuation e.g. ` much.]`
- probably leave internal punctuation untouched e.g. "its" and "it's" should still be counted separately

I think the naive approach would be enough (or better).

One drawback of changing the word count metric calculation is that when it changes, people will suddenly see a drop in their total word count, revised word count, and percentage of 10,000 words, which might be disheartening (in the short term, when it first changes). This would likely be more noticeable the lower the count (e.g. when you're just starting out). 

(You might happen to see an increase in memorised words as some words would be re-distributed from revised words to memorised words when combined.)

We might add a new disclaimer like `<p class="text-small">Note: a “word” is a unique word, ignoring some punctuation and spacing.</p>` or whatever we decide on.

What are your thoughts @jladdjr ?

**Comment by jladdjr** [comment]
TLDR - basically, everything makes sense / sounds good 👍 . Appreciate you confirming the current behavior and fleshing out what the changes would entail.

----
The slightly longer response..

Thanks for going over the current behavior. Makes sense.

> This means the progress shown is an accurate measure but perhaps not as obvious/learnable or satisfying.

Also makes sense. 

> We could update the 6 Progress page word count metrics to add together the counts for entries containing the same words with and without punctuation.

Sounds good; as I've continued to think about this, this still seems like it'd be a helpful change.

> We may or may not update the 6 Progress page word count metrics to add together the counts for entries with different leading/trailing spacing.

Given that most folks would stick with one setting for spacing - either printing the space before or after - this does seem less important than punctuation. 👍 

Re: naive vs. sophisticated approach, makes sense that it'd be tricky to create rules that successfully cover _every_ possible case. I agree that the naive approach sounds totally fine. It'd cover the lion's share of cases.

> One drawback of changing the word count metric calculation is that when it changes, people will suddenly see a drop in their total word count.. You might happen to see an increase in memorised words.

Yep, makes sense. Adding the notice that you mentioned sounds like a good idea. Think the change would still be a good one on balance, despite the hiccup.

**Comment by didoesdigital** [comment]
Ok, let's

1. Update the 6 Progress page word count metrics to add together the counts for entries containing the same words with and without punctuation
2. Update the progress revision lessons setup to match
3. Use the naive approach to decide what punctuation counts as "the same": trim trailing commas, full stops, exclamation marks, and question marks before adding its count to the word's count.

@jladdjr are you interested in working on this? If so we can assign this issue to you.

**Comment by jladdjr** [comment]
@didoesdigital Definitely interested, but pretty short on cycles at the moment (just had baby #2 join our family recently) and also a level zero javascript n00b. Would love to dig into the code at some point, but not sure that I could do this in anything resembling a timely manner.

**Comment by didoesdigital** [comment]
Oh, congrats @jladdjr ! 

And no worries, we'll leave the issue unassigned.
